sources/bloodhound/src/main.rs

```diff
@@ -0,0 +1,212 @@
+  -l, --level <LEVEL>       The CIS benchmark compliance level to check [default: 1]
+  -o, --output <OUTPUT>     Write output to a file at given path [default: stdout]
+  -h, --help                Print help
+```
+*/
+
+use bloodhound::args::*;
+use bloodhound::output::{JsonReportWriter, ReportWriter, TextReportWriter};
+use bloodhound::results::{
+    CheckStatus, CheckerMetadata, CheckerResult, ReportMetadata, ReportResults,
+};
+use std::collections::HashMap;
+use std::fs::{DirEntry, File};
+use std::io::{stdout, BufReader, Error, Write};
+use std::os::unix::fs::PermissionsExt;
+use std::path::Path;
+use std::process::{Command, Output};
+use std::{fs, path::PathBuf};
+
+// Define some exit codes for error conditions
+const CHECKER_DISCOVERY_ERROR: i32 = 2;
+const REPORT_OUTPUT_ERROR: i32 = 3;
+const NO_CHECKS_RUN_ERROR: i32 = 4;
+
+/// Discovers the metadata information for the checkers being run or provides a default.
+fn read_metadata(check_dir: &Path) -> ReportMetadata {
+    let meta_path = check_dir.join("metadata.json");
+
+    if let Ok(file) = File::open(meta_path) {
+        let reader = BufReader::new(file);
+        if let Ok(report_metadata) = serde_json::from_reader(reader) {
+            return report_metadata;
+        }
+    }
+
+    ReportMetadata {
+        name: None,
+        version: None,
+        url: None,
+    }
+}
+
+/// Discover all executable checker files for the given directory.
+/// By looking at the provided path, this does some basic checks and filtering
+/// to find the executable files that appear to be bloodhound checkers.
+/// It will parse the checker metadata to provide a mapping between a filesystem
+/// path and a specific check, including only those checkers that are within the
+/// request compliance level.
+fn find_checkers(check_dir: &PathBuf, level: u8) -> HashMap<String, CheckerMetadata> {
+    let entries: Vec<DirEntry> = fs::read_dir(check_dir)
+        .unwrap()
+        .filter_map(|file| file.ok())
+        .collect();
+    let mut result = HashMap::new();
+
+    for entry in entries {
+        if let Ok(file_metadata) = fs::metadata(entry.path()) {
+            // Filter out any subdirectories
+            if file_metadata.is_dir() {
+                continue;
+            }
+
+            // Skip any files that are not executable
+            if file_metadata.permissions().mode() & 0o111 == 0 {
+                continue;
+            }
+
+            // It's an executable file, make sure it implements our expected checker interface
+            let metadata;
+            if let Ok(output) = Command::new(entry.path()).arg("metadata").output() {
+                metadata = String::from_utf8_lossy(&output.stdout).to_string();
+            } else {
+                eprintln!(
+                    "{:?} does not appear to be a checker executable",
+                    entry.path()
+                );
+                continue;
+            }
```

Comment:
Line 106 handles the second and third bullet items. I can move this into a separate helper function, but I wanted to have it here since it is a small amount of logic and easier to read through and see what's happening.